{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "from aerobot.io import RESULTS_PATH\n",
    "from aerobot.dataset import dataset_load_feature_order\n",
    "from aerobot.kmer import kmer_sequence_to_kmers\n",
    "import os \n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Black Sea genome data is in the form of nucleotide contigs. I ran prodigal on the raw nucleotide data (using `prodigal -i bs_contigs.fna -o bs_contigs.gbk -a bs_contigs.faa` ), which resulted in a FASTA nucleotide of amino acid sequences. Each header string contains a unique ID, e.g. GCA_014383885.1.1_3, where the first number following the genome ID (GCA_014383885.1) is the contig from which the predicted amino acid sequence was produced. \n",
    "\n",
    "The goal is to re-train an amino acid trimer-based classifier using only the amino acids from contigs. I can then compare the results to the performance of the whole-genome classifier by mapping the predictions from each contig to the predictions of the genome to which the contig belongs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "BS_PATH = os.path.join(RESULTS_PATH, 'black_sea')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "aa_3mer_features = dataset_load_feature_order('aa_3mer')\n",
    "\n",
    "# Load in the contig k-mers, which were generated by running the count-kmers.py script on HPC. \n",
    "kmer_df = pd.read_csv(os.path.join(BS_PATH, 'bs_aa_3mer_from_contigs.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K-mer counts generated for 54544 contigs.\n"
     ]
    }
   ],
   "source": [
    "print(f'K-mer counts generated for {len(kmer_df)} contigs.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "# The k-mer counts seem pretty sparse. In order to make sure everything worked correctly, I am going to merge the contig k-mer counts\n",
    "# by genome, and compare them to the k-mer counts in the bs_aa_3mer.csv file (which Josh generated).\n",
    "\n",
    "def get_genome_id(contig_id:str):\n",
    "    '''Get the genome ID from the contig ID, which is of the form {genome_id}.{contig_index}.'''\n",
    "    match = re.search('(GCA_\\d+\\.\\d+)\\.\\d+', contig_id)\n",
    "    return match.group(1) # Extract the genome ID from the pattern and return. \n",
    "\n",
    "kmer_df['genome_id'] = kmer_df.contig_id.apply(get_genome_id)\n",
    "kmer_df_by_genome = kmer_df.drop(columns=['contig_id']).groupby('genome_id').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in the original k-mers. \n",
    "kmer_df_by_genome_ref = pd.read_csv(os.path.join(BS_PATH, 'bs_aa_3mer.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardize the column order and k-mer counts included in each DataFrame. \n",
    "kmer_df_by_genome = kmer_df_by_genome[aa_3mer_features]\n",
    "kmer_df_by_genome_ref = kmer_df_by_genome_ref[aa_3mer_features]\n",
    "# Also make sure the index order matches for easier comparison. \n",
    "kmer_df_by_genome, kmer_df_by_genome_ref = kmer_df_by_genome.align(kmer_df_by_genome_ref, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTA</th>\n",
       "      <th>TAT</th>\n",
       "      <th>ATG</th>\n",
       "      <th>TGG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GTW</th>\n",
       "      <th>TWS</th>\n",
       "      <th>WSI</th>\n",
       "      <th>SID</th>\n",
       "      <th>IDL</th>\n",
       "      <th>...</th>\n",
       "      <th>QCC</th>\n",
       "      <th>MYC</th>\n",
       "      <th>CQC</th>\n",
       "      <th>NCC</th>\n",
       "      <th>MYW</th>\n",
       "      <th>MWC</th>\n",
       "      <th>CMC</th>\n",
       "      <th>CNW</th>\n",
       "      <th>CCM</th>\n",
       "      <th>CWC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_014381205.1</th>\n",
       "      <td>117.0</td>\n",
       "      <td>143.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381215.1</th>\n",
       "      <td>125.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>191.0</td>\n",
       "      <td>218.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381225.1</th>\n",
       "      <td>128.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>241.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381265.1</th>\n",
       "      <td>91.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>205.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381285.1</th>\n",
       "      <td>107.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>134.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 8000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   TTA    TAT    ATG    TGG    GGT   GTW   TWS   WSI    SID  \\\n",
       "GCA_014381205.1  117.0  143.0  221.0  222.0  216.0  30.0  31.0  18.0  102.0   \n",
       "GCA_014381215.1  125.0  141.0  171.0  191.0  218.0  49.0  35.0  30.0  156.0   \n",
       "GCA_014381225.1  128.0  146.0  302.0  220.0  241.0  21.0  11.0  13.0   95.0   \n",
       "GCA_014381265.1   91.0   93.0   86.0  109.0   99.0  28.0  18.0  24.0  205.0   \n",
       "GCA_014381285.1  107.0  128.0  118.0  134.0  153.0  12.0  18.0  19.0  196.0   \n",
       "\n",
       "                   IDL  ...  QCC  MYC  CQC  NCC   MYW  MWC  CMC  CNW  CCM  CWC  \n",
       "GCA_014381205.1  160.0  ...  2.0  3.0  0.0  0.0   1.0  3.0  1.0  2.0  1.0  0.0  \n",
       "GCA_014381215.1  230.0  ...  1.0  3.0  0.0  5.0  11.0  0.0  2.0  3.0  2.0  2.0  \n",
       "GCA_014381225.1  195.0  ...  2.0  3.0  2.0  0.0   1.0  0.0  0.0  2.0  0.0  1.0  \n",
       "GCA_014381265.1  236.0  ...  4.0  3.0  1.0  5.0   2.0  0.0  2.0  8.0  1.0  0.0  \n",
       "GCA_014381285.1  228.0  ...  1.0  2.0  2.0  6.0   7.0  4.0  1.0  2.0  2.0  0.0  \n",
       "\n",
       "[5 rows x 8000 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_df_by_genome.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTA</th>\n",
       "      <th>TAT</th>\n",
       "      <th>ATG</th>\n",
       "      <th>TGG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GTW</th>\n",
       "      <th>TWS</th>\n",
       "      <th>WSI</th>\n",
       "      <th>SID</th>\n",
       "      <th>IDL</th>\n",
       "      <th>...</th>\n",
       "      <th>QCC</th>\n",
       "      <th>MYC</th>\n",
       "      <th>CQC</th>\n",
       "      <th>NCC</th>\n",
       "      <th>MYW</th>\n",
       "      <th>MWC</th>\n",
       "      <th>CMC</th>\n",
       "      <th>CNW</th>\n",
       "      <th>CCM</th>\n",
       "      <th>CWC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_014381205.1</th>\n",
       "      <td>135.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>215.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>213.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>173.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381215.1</th>\n",
       "      <td>124.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>193.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>223.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381225.1</th>\n",
       "      <td>129.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>193.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381265.1</th>\n",
       "      <td>88.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>199.0</td>\n",
       "      <td>231.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_014381285.1</th>\n",
       "      <td>106.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>191.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 8000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   TTA    TAT    ATG    TGG    GGT   GTW   TWS   WSI    SID  \\\n",
       "GCA_014381205.1  135.0  162.0  215.0  217.0  213.0  37.0  34.0  19.0  102.0   \n",
       "GCA_014381215.1  124.0  141.0  170.0  193.0  216.0  47.0  36.0  30.0  154.0   \n",
       "GCA_014381225.1  129.0  151.0  294.0  219.0  239.0  21.0  12.0  14.0   95.0   \n",
       "GCA_014381265.1   88.0   91.0   83.0  103.0   98.0  28.0  18.0  24.0  199.0   \n",
       "GCA_014381285.1  106.0  120.0  117.0  129.0  149.0  12.0  17.0  19.0  191.0   \n",
       "\n",
       "                   IDL  ...  QCC  MYC  CQC  NCC   MYW  MWC  CMC  CNW  CCM  CWC  \n",
       "GCA_014381205.1  173.0  ...  2.0  4.0  0.0  1.0   1.0  2.0  1.0  1.0  2.0  0.0  \n",
       "GCA_014381215.1  223.0  ...  1.0  3.0  0.0  6.0  11.0  0.0  2.0  3.0  2.0  2.0  \n",
       "GCA_014381225.1  193.0  ...  2.0  3.0  2.0  0.0   2.0  0.0  1.0  2.0  0.0  1.0  \n",
       "GCA_014381265.1  231.0  ...  3.0  3.0  1.0  5.0   2.0  0.0  2.0  8.0  0.0  0.0  \n",
       "GCA_014381285.1  221.0  ...  0.0  2.0  2.0  6.0   7.0  4.0  1.0  2.0  1.0  0.0  \n",
       "\n",
       "[5 rows x 8000 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Counts in the reference and contig-based k-mer DataFrames seem very similar, though not identical. I am pretty sure Prodigal is \n",
    "# deterministic, so perhaps Josh used a different set of contigs, or erged the contig sequences before counting k-mers. I \n",
    "# will note that it seems as though the counts in the reference are consistently higher than the counts I generated, so that\n",
    "# explanation is plausible. \n",
    "kmer_df_by_genome_ref.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have trained amino acid trimer-based models (both logistic and nonlinear) saved in the notebooks directory. These can be used to generate predictions for the contig trimers. Predictions needed to be generated on the HPC (they require too much memory to run locally). They were produced using the commands below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! python ../scripts/run.py 'nonlinear' --feature-type 'aa_3mer' --save-model 1 --n-epochs 500 --save-model-path 'aa_3mer_nonlinear_model.joblib'\n",
    "# ! python ../scripts/run.py 'logistic' --feature-type 'aa_3mer' --save-model 1 --max-iter 10000 --save-model-path 'aa_3mer_logistic_model.joblib'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 aa_3mer features are missing from the input data. Filled missing data with 0.\n"
     ]
    }
   ],
   "source": [
    "# TODO: Create a models directory to store pre-trained models for better organization. \n",
    "\n",
    "# ! python ../scripts/run-pretrained.py -i ../../results/black_sea/bs_aa_3mer_from_contigs.csv -o ../../results/black_sea/bs_run_pretrained_results_logistic_from_contigs.csv -m aa_3mer_logistic.joblib\n",
    "# ! python ../scripts/run-pretrained.py -i ../../results/black_sea/bs_aa_3mer_from_contigs.csv -o ../../results/black_sea/bs_run_pretrained_results_nonlinear_from_contigs.csv -m aa_3mer_nonlinear.joblib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aerobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
