{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from aerobot.utils import FEATURE_TYPES, DATA_PATH, MODELS_PATH, save_hdf, RNA16S_PATH\n",
    "from aerobot.models import NonlinearClassifier\n",
    "from aerobot.dataset import FeatureDataset\n",
    "from aerobot import ncbi\n",
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import wget\n",
    "import zipfile\n",
    "import warnings\n",
    "import tables\n",
    "import shutil\n",
    "\n",
    "# Ignore some annoying warnings triggered when saving HDF files.\n",
    "warnings.filterwarnings('ignore', category=pd.io.pytables.PerformanceWarning)\n",
    "warnings.filterwarnings('ignore', category=tables.NaturalNameWarning)\n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of RNA 16S training dataset: 693\n",
      "Size of RNA 16S testing dataset: 188\n",
      "Size of RNA 16S validation dataset: 150\n",
      "\n",
      "Total number of 16S RNA sequences: 1031\n"
     ]
    }
   ],
   "source": [
    "rna16s_training_dataset = FeatureDataset(os.path.join(RNA16S_PATH, 'training_dataset.h5'), feature_type='embedding_rna16s')\n",
    "rna16s_testing_dataset = FeatureDataset(os.path.join(RNA16S_PATH, 'testing_dataset.h5'), feature_type='embedding_rna16s')\n",
    "rna16s_validation_dataset = FeatureDataset(os.path.join(RNA16S_PATH, 'validation_dataset.h5'), feature_type='embedding_rna16s')\n",
    "\n",
    "print('Size of RNA 16S training dataset:', len(rna16s_training_dataset))\n",
    "print('Size of RNA 16S testing dataset:', len(rna16s_testing_dataset))\n",
    "print('Size of RNA 16S validation dataset:', len(rna16s_validation_dataset))\n",
    "print()\n",
    "print('Total number of 16S RNA sequences:', len(rna16s_training_dataset) + len(rna16s_testing_dataset) + len(rna16s_validation_dataset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/prichter/Documents/aerobot/aerobot/../data/suppressed_genomes.tsv'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the list of suppressed genomes from the NCBI FTP site. \n",
    "# These genomes did not meet these standards: https://www.ncbi.nlm.nih.gov/datasets/docs/v2/policies-annotation/genome-processing/genome_notes/ \n",
    "wget.download('ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq_historical.txt', os.path.join(DATA_PATH, 'suppressed_genomes.tsv'))\n",
    "suppressed_genomes_df = pd.read_csv(os.path.join(DATA_PATH, 'suppressed_genomes.tsv'), delimiter='\\t', skiprows=1, index_col=0)\n",
    "suppressed_genomes_df.index.name = 'id'\n",
    "suppressed_genomes_df = suppressed_genomes_df[['version_status', 'organism_name', 'annotation_provider', 'annotation_date']]\n",
    "suppressed_genomes_df.to_csv(os.path.join(DATA_PATH, 'suppressed_genomes.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting tired of contending with all the filename mapping, so just going to go ahead and adjust the source data.\n",
    "\n",
    "old_madin_data = pd.HDFStore(os.path.join(DATA_PATH, 'original', 'madin.h5'))\n",
    "new_madin_data = dict()\n",
    "\n",
    "# Split the things which were previously stored as metadata into their individual features. \n",
    "old_metadata = old_madin_data['AF'].rename(columns={'oxygen_genes':'number_of_oxygen_genes'}).drop_duplicates()\n",
    "old_metadata['percent_oxygen_genes'] = old_metadata['number_of_oxygen_genes'] / old_metadata['number_of_genes']\n",
    "new_madin_data['number_of_genes'] = old_metadata[['number_of_genes']] \n",
    "new_madin_data['number_of_oxygen_genes'] = old_metadata[['number_of_oxygen_genes']] \n",
    "new_madin_data['percent_oxygen_genes'] = old_metadata[['percent_oxygen_genes']] \n",
    "\n",
    "# Store what used to be called \"labels\" under \"metadata.\" Drop some of the long, annoying columns.\n",
    "new_madin_data['metadata'] = old_madin_data['labels'].drop(columns=['annotation_file', 'embedding_file']).drop_duplicates()\n",
    "new_madin_data['ko'] = old_madin_data['KO']\n",
    "new_madin_data['embedding_genome'] = old_madin_data['WGE']\n",
    "new_madin_data['embedding_oxygen_genes'] = old_madin_data['OGSE']\n",
    "\n",
    "exclude_feature_types = ['ko_terminal_oxidase_genes', 'chemical', 'embedding_rna16s']\n",
    "for feature_type in FEATURE_TYPES:\n",
    "    if (feature_type not in new_madin_data) and (feature_type not in exclude_feature_types):\n",
    "        new_madin_data[feature_type] = old_madin_data[feature_type]\n",
    "\n",
    "save_hdf(new_madin_data, os.path.join(DATA_PATH, 'madin_datasets.h5'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Also going to put the Jablonska data into an HDF file for consistency. \n",
    "new_jablonska_data = dict()\n",
    "\n",
    "# Split the things which were previously stored as metadata into their individual features. \n",
    "old_metadata =  pd.read_csv(os.path.join(DATA_PATH, 'original', f'jablonska_metadata.csv'), index_col=0)\n",
    "old_metadata = old_metadata.rename(columns={'oxygen_genes':'number_of_oxygen_genes'}).drop_duplicates()\n",
    "old_metadata = old_metadata.set_index('genome')\n",
    "old_metadata['percent_oxygen_genes'] = old_metadata['number_of_oxygen_genes'] / old_metadata['number_of_genes']\n",
    "old_metadata = old_metadata.rename({'Oder':'Order'})\n",
    "\n",
    "new_jablonska_data['number_of_genes'] = old_metadata[['number_of_genes']] \n",
    "new_jablonska_data['number_of_oxygen_genes'] = old_metadata[['number_of_oxygen_genes']] \n",
    "new_jablonska_data['percent_oxygen_genes'] = old_metadata[['percent_oxygen_genes']] \n",
    "\n",
    "# Store what used to be called \"labels\" under \"metadata.\" Drop some of the long, annoying columns.\n",
    "new_metadata = pd.read_csv(os.path.join(DATA_PATH, 'original', f'jablonska_labels.csv'), index_col=0).drop(columns=['annotation_file', 'embedding_file']).drop_duplicates()\n",
    "new_metadata = new_metadata.rename(columns={'Oder':'Order'})\n",
    "new_jablonska_data['metadata'] = new_metadata\n",
    "\n",
    "exclude_feature_types = ['ko_terminal_oxidase_genes', 'chemical', 'embedding_rna16s']\n",
    "for feature_type in FEATURE_TYPES:\n",
    "    if (feature_type not in new_jablonska_data) and (feature_type not in exclude_feature_types):\n",
    "        new_jablonska_data[feature_type] = pd.read_csv(os.path.join(DATA_PATH, 'original', f'jablonska_{feature_type}.csv'), index_col=0)\n",
    "\n",
    "save_hdf(new_jablonska_data, os.path.join(DATA_PATH, 'jablonska_datasets.h5'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Curious about how much the annotated KO groups in the Jablonska and Madin datasets overlapped. I am running into a bug due to the fact that one (or both) of the datasets have no proteins annotated with some of the KO groups in the list of terminal oxidase KO groups. I am trying to figure out if it is better to fill in zeros or drop the missing KO groups.\n",
    "\n",
    "If both datasets are missing the same KO groups, I will want to drop the KO groups which are not present. If I fill in with zeros, then the model weights corresponding to those input features will not be trained. If input data which *does* contain those KO groups is used as input to the trained model, then it could adversely effect predictions.\n",
    "\n",
    "I first want to make sure that the columns in the Jablonska and Madin KO data are the same. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000010085.1 GCA_000069185.1 GCA_000093085.1 GCA_002339675.1 GCA_002363385.1 GCA_002386405.1 GCA_003165155.1 GCA_003241595.1 GCA_003248565.1 GCA_003248865.1 GCA_003447235.1 GCA_003483565.1 GCA_003510825.1 GCA_003694355.1 GCA_004298145.1 GCA_004554205.1 GCA_004796465.1 GCA_004799495.1 GCA_005779435.1 GCA_009886055.1 GCA_009886125.1 GCA_011524565.1 GCA_012270465.1 GCA_012523755.1 GCA_012837295.1 GCA_013361785.1 GCA_014376935.1 GCA_014801735.1 GCA_014801785.1 GCA_014801815.1 GCA_014801865.1 GCA_014801985.1 GCA_014802105.1 GCA_014802425.1 GCA_014803725.1 GCA_014804865.1 GCA_015058375.1 GCA_015255025.1 GCA_016177775.1 GCA_016208105.1 GCA_016281815.1 GCA_016284635.1 GCA_016286585.1 GCA_016287755.1 GCA_016291915.1 GCA_016703745.1 GCA_016710135.1 GCA_016841205.1 GCA_017304655.1 GCA_017402445.1 GCA_017423565.1 GCA_017435645.1 GCA_017449675.1 GCA_017458005.1 GCA_017459445.1 GCA_017460515.1 GCA_017467565.1 GCA_017467705.1 GCA_017469955.1 GCA_017479675.1 GCA_017483185.1 GCA_017538195.1 GCA_017547605.1 GCA_017548605.1 GCA_017619375.1 GCA_017635615.1 GCA_017645905.1 GCA_017879645.1 GCA_017994575.1 GCA_018054505.1 GCA_018065145.1 GCA_018260395.1 GCA_018262705.1 GCA_018366675.1 GCA_018371855.1 GCA_018375675.1 GCA_018882965.1 GCA_019242285.1 GCA_019246375.1 GCA_900444995.1 GCA_900696665.1 GCA_902512305.1 GCA_902513495.1 GCA_902535975.1 GCA_902563055.1 GCA_902826385.1 GCA_902826455.1 GCA_903832295.1 GCA_903837345.1 GCA_903844415.1 GCA_903862645.1 GCA_903866355.1 GCA_903870185.1 GCA_903938575.1 GCA_905182855.1 GCA_905479545.1 GCA_905479875.1 GCF_000007825.1 GCF_000008465.1 GCF_000009045.1 GCF_000010985.1 GCF_000011085.1 GCF_000011205.1 GCF_000011305.1 GCF_000013345.1 GCF_000013565.1 GCF_000014005.1 GCF_000014805.1 GCF_000015025.1 GCF_000015045.1 GCF_000015125.1 GCF_000015285.1 GCF_000015725.1 GCF_000015805.1 GCF_000016125.1 GCF_000016165.1 GCF_000017185.1 GCF_000017265.1 GCF_000017405.1 GCF_000017625.1 GCF_000018025.1 GCF_000018205.1 GCF_000018285.1 GCF_000018365.1 GCF_000018685.1 GCF_000018945.1 GCF_000021485.1 GCF_000021685.1 GCF_000021745.1 GCF_000023105.1 GCF_000023125.1 GCF_000023245.1 GCF_000023265.1 GCF_000023585.1 GCF_000024025.1 GCF_000024065.1 GCF_000024105.1 GCF_000024425.1 GCF_000025605.1 GCF_000026145.1 GCF_000026185.1 GCF_000027165.1 GCF_000058485.1 GCF_000062885.1 GCF_000069225.1 GCF_000091545.1 GCF_000092245.1 GCF_000092465.1 GCF_000092505.1 GCF_000143145.1 GCF_000143685.1 GCF_000143765.1 GCF_000144625.1 GCF_000146585.1 GCF_000147075.1 GCF_000152905.1 GCF_000153285.1 GCF_000153305.1 GCF_000153465.1 GCF_000153885.1 GCF_000153905.1 GCF_000154465.1 GCF_000154845.1 GCF_000155595.1 GCF_000155775.1 GCF_000157995.1 GCF_000159315.1 GCF_000159695.1 GCF_000160055.1 GCF_000160615.1 GCF_000162875.1 GCF_000166275.1 GCF_000166295.1 GCF_000166335.1 GCF_000169235.1 GCF_000172975.1 GCF_000173355.1 GCF_000173655.1 GCF_000173735.1 GCF_000174755.1 GCF_000175195.1 GCF_000176015.1 GCF_000176175.1 GCF_000176235.1 GCF_000178455.1 GCF_000179055.1 GCF_000179555.1 GCF_000183385.1 GCF_000184685.1 GCF_000184745.1 GCF_000185905.1 GCF_000186265.1 GCF_000187585.1 GCF_000188175.1 GCF_000190315.1 GCF_000190635.1 GCF_000190735.1 GCF_000194605.1 GCF_000196235.1 GCF_000196895.1 GCF_000200735.1 GCF_000203935.1 GCF_000210715.1 GCF_000212815.1 GCF_000213055.1 GCF_000215485.1 GCF_000217935.1 GCF_000218895.1 GCF_000219585.1 GCF_000220945.1 GCF_000222305.1 GCF_000223395.1 GCF_000224085.1 GCF_000224785.1 GCF_000225705.1 GCF_000226875.1 GCF_000227275.1 GCF_000227465.1 GCF_000227765.1 GCF_000230675.1 GCF_000230835.1 GCF_000233455.1 GCF_000233955.1 GCF_000235445.1 GCF_000235505.1 GCF_000236865.1 GCF_000237085.1 GCF_000238795.1 GCF_000242235.1 GCF_000243235.1 GCF_000244915.1 GCF_000245555.1 GCF_000250635.1 GCF_000252995.1 GCF_000255555.1 GCF_000262465.1 GCF_000263635.1 GCF_000265385.1 GCF_000265505.1 GCF_000273375.1 GCF_000277125.1 GCF_000281895.1 GCF_000284295.1 GCF_000286075.1 GCF_000297495.1 GCF_000298295.1 GCF_000307165.1 GCF_000311785.1 GCF_000312005.1 GCF_000316665.1 GCF_000317635.1 GCF_000317675.1 GCF_000327505.1 GCF_000331575.1 GCF_000332155.1 GCF_000334215.1 GCF_000334405.1 GCF_000336615.1 GCF_000336775.1 GCF_000336815.1 GCF_000337195.1 GCF_000337455.1 GCF_000337735.1 GCF_000341345.1 GCF_000347635.1 GCF_000353285.1 GCF_000364625.1 GCF_000368025.1 GCF_000372085.1 GCF_000372405.1 GCF_000373245.1 GCF_000373765.1 GCF_000373965.1 GCF_000374025.1 GCF_000374465.1 GCF_000375365.1 GCF_000375405.1 GCF_000376805.1 GCF_000377005.1 GCF_000377505.1 GCF_000377665.1 GCF_000377705.1 GCF_000378085.1 GCF_000378205.1 GCF_000378345.1 GCF_000378445.1 GCF_000378945.1 GCF_000379805.1 GCF_000379945.1 GCF_000380125.1 GCF_000380145.1 GCF_000382125.1 GCF_000384195.1 GCF_000384215.1 GCF_000400735.1 GCF_000406925.1 GCF_000417505.1 GCF_000420325.1 GCF_000420765.1 GCF_000421265.1 GCF_000421565.1 GCF_000421825.1 GCF_000422345.1 GCF_000422565.1 GCF_000422585.1 GCF_000423185.1 GCF_000423325.1 GCF_000423425.1 GCF_000423465.1 GCF_000423525.1 GCF_000423745.1 GCF_000424665.1 GCF_000425005.1 GCF_000425165.1 GCF_000425485.1 GCF_000425545.1 GCF_000425825.1 GCF_000426665.1 GCF_000427385.1 GCF_000427645.1 GCF_000427725.1 GCF_000427965.1 GCF_000429245.1 GCF_000429325.1 GCF_000429525.1 GCF_000429705.1 GCF_000430545.1 GCF_000430585.1 GCF_000430665.1 GCF_000430845.1 GCF_000455445.1 GCF_000468055.1 GCF_000472545.1 GCF_000472705.1 GCF_000482505.1 GCF_000504285.1 GCF_000508085.1 GCF_000512355.1 GCF_000512735.1 GCF_000519125.1 GCF_000524555.1 GCF_000568815.1 GCF_000585355.1 GCF_000600005.1 GCF_000612665.1 GCF_000613345.1 GCF_000615285.1 GCF_000619845.1 GCF_000620105.1 GCF_000620985.1 GCF_000621085.1 GCF_000621125.1 GCF_000621145.1 GCF_000661915.1 GCF_000685805.1 GCF_000686605.1 GCF_000686725.1 GCF_000688295.1 GCF_000688515.1 GCF_000701465.1 GCF_000711735.1 GCF_000711895.1 GCF_000733855.1 GCF_000735365.1 GCF_000738005.1 GCF_000739085.1 GCF_000740675.1 GCF_000745485.1 GCF_000745955.1 GCF_000747155.1 GCF_000775615.1 GCF_000785515.1 GCF_000829035.1 GCF_000829825.1 GCF_000960085.1 GCF_000969885.1 GCF_000973105.1 GCF_001025195.1 GCF_001027105.1 GCF_001038625.1 GCF_001061665.1 GCF_001074805.1 GCF_001075875.1 GCF_001078705.1 GCF_001180165.1 GCF_001185345.1 GCF_001263795.1 GCF_001274595.1 GCF_001281025.1 GCF_001295365.1 GCF_001308265.1 GCF_001310815.1 GCF_001314995.1 GCF_001436305.1 GCF_001476815.1 GCF_001477655.1 GCF_001578705.1 GCF_001578935.1 GCF_001588645.1 GCF_001602525.1 GCF_001624705.1 GCF_001653475.1 GCF_001662485.1 GCF_001693735.1 GCF_001885235.1 GCF_001908725.1 GCF_001941965.1 GCF_001953955.1 GCF_002078505.1 GCF_002093975.1 GCF_002094795.1 GCF_002095385.1 GCF_002095465.1 GCF_002096845.1 GCF_002105465.1 GCF_002115725.1 GCF_002200015.1 GCF_002206625.1 GCF_002240055.1 GCF_002494345.1 GCF_002744245.1 GCF_002778015.1 GCF_002797535.1 GCF_002797685.1 GCF_002797755.1 GCF_002797855.1 GCF_002799165.1 GCF_002849695.1 GCF_002866065.1 GCF_002890795.1 GCF_002917105.1 GCF_002917885.1 GCF_002924625.1 GCF_002934625.1 GCF_002939475.1 GCF_002960425.1 GCF_003002095.1 GCF_003014485.1 GCF_003014755.1 GCF_003030305.1 GCF_003046225.1 GCF_003054175.1 GCF_003054225.1 GCF_003054245.1 GCF_003148495.1 GCF_003148745.1 GCF_003201285.1 GCF_003201875.1 GCF_003202695.1 GCF_003315205.1 GCF_003326795.1 GCF_003337215.1 GCF_003349955.1 GCF_003351105.1 GCF_003634095.1 GCF_003634455.1 GCF_003671955.1 GCF_003688405.1 GCF_003688935.1 GCF_003725815.1 GCF_003732425.1 GCF_003752135.1 GCF_003752415.1 GCF_003752585.1 GCF_003814915.1 GCF_003942585.1 GCF_003942695.1 GCF_003943395.1 GCF_003943815.1 GCF_003945385.1 GCF_003970865.1 GCF_004054105.1 GCF_004339595.1 GCF_004340985.1 GCF_004341235.1 GCF_004345265.1 GCF_004346685.1 GCF_004346865.1 GCF_004362405.1 GCF_004363295.1 GCF_004363305.1 GCF_004364935.1 GCF_004364975.1 GCF_004684085.1 GCF_004802395.1 GCF_005144795.1 GCF_005843985.1 GCF_005844455.1 GCF_005886145.1 GCF_006539885.1 GCF_006716675.1 GCF_006970445.1 GCF_008327825.1 GCF_008639165.1 GCF_009689985.1 GCF_009696605.1 GCF_011758765.1 GCF_012273215.1 GCF_012932795.1 GCF_012974785.1 GCF_013186535.1 GCF_013389365.1 GCF_013410345.1 GCF_013464915.1 GCF_014191425.1 GCF_014192415.1 GCF_014192515.1 GCF_014199535.1 GCF_014204915.1 GCF_015561235.1 GCF_015709225.1 GCF_016028255.1 GCF_016648925.1 GCF_016658865.1 GCF_016728505.1 GCF_016775005.1 GCF_016901815.1 GCF_017299535.1 GCF_018127965.1 GCF_018129995.1 GCF_018141405.1 GCF_018223745.1 GCF_018228765.1 GCF_018502625.1 GCF_018614955.1 GCF_019212185.1 GCF_900107675.1 GCF_900109575.1 GCF_900109705.1 GCF_900111145.1 GCF_900114615.1 GCF_900115845.1 GCF_900119315.1 GCF_900128975.1 GCF_900129055.1 GCF_900130005.1 GCF_900141635.1 GCF_900141805.1 GCF_900141815.1 GCF_900142625.1 GCF_900142785.1 GCF_900143645.1 GCF_900156045.1 GCF_900167125.1 GCF_900172255.1 GCF_900172285.1 GCF_900177295.1 GCF_900182705.1 GCF_900182725.1 GCF_900187995.1 GCF_900188035.1 GCF_900188065.1 GCF_900215245.1 GCF_900290975.1 GCF_900445025.1 GCF_900445265.1 GCF_900447045.1 GCF_900636475.1 GCF_900636825.1 GCF_900636965.1 GCF_900637905.1 GCF_900638585.1 GCF_902497895.1 GCF_902497995.1 GCF_902498135.1 GCF_902983995.1 GCF_904846225.1 GCF_000008485.1 GCF_000008285.1 GCF_000219725.1 GCF_000007625.1 GCF_000195185.1 GCF_000009005.1 GCF_000284355.1 GCF_000192315.1 GCF_000193205.1 GCF_000010125.1 GCF_000006845.1 GCF_000020725.1 GCF_000046705.1 GCF_000252975.1 GCF_000022785.1 GCF_000008665.1 GCF_000010585.1 GCF_000008505.1 GCF_000015485.1 GCF_000020625.1\n"
     ]
    }
   ],
   "source": [
    "testing_genomes = pd.read_hdf(os.path.join(DATA_PATH, 'testing_datasets.h5'), key='metadata').index\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seems like a best call to use the union of the columns in each dataset. This should maximize the amount of information, while ensuring that all weights are actually updated during model training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code I ran to clean up the Westoby data and add taxonomy. \n",
    "\n",
    "metadata = pd.read_csv(os.path.join(DATA_PATH, 'Mark_Westoby_Organism_Metadata_Export_02152018.tsv'), sep='\\t', index_col=0).GENBANK_16S_ID\n",
    "print(metadata)\n",
    "metadata = metadata[['GENBANK_16S_ID','OXYGEN_REQUIREMENT', 'NCBI_TAXONOMY_ID']] # Grab the relevant columns from the metadata DataFrame.\n",
    "\n",
    "metadata = metadata[~metadata['GENBANK_16S_ID'].str.contains('(null)', regex=False)]\n",
    "metadata = metadata[~metadata['GENBANK_16S_ID'].str.contains('(null)', regex=False)]\n",
    "metadata.columns = ['id', 'physiology', 'ncbi_taxonomy_id']\n",
    "\n",
    "# Some of the IDs have multiple GenBank accessions listed for the same organism (comma-separated)\n",
    "# I assume these are duplicates, and just used the first one. \n",
    "metadata['id'] = [id_.split(',')[0] for id_ in metadata['id']] \n",
    "# Other IDs seem to have extra information appended to the end, following a colon. I assume I can remove this. \n",
    "metadata['id'] = [id_.split(':')[0] for id_ in metadata['id']]\n",
    "\n",
    "label_map = {'Anaerobe':'anaerobe', 'Aerobe':'aerobe', 'Facultative':'facultative', 'Obligate anaerobe':'anaerobe', 'Obligate aerobe':'aerobe','Facultative anaerobe':'facultative'}\n",
    "metadata = metadata[metadata.physiology.isin(label_map)] # Filter for organisms with an oxygen label is in the map. \n",
    "metadata.physiology = metadata.physiology.replace(label_map)\n",
    "metadata = metadata.drop_duplicates('id')\n",
    "metadata = metadata.set_index('id')\n",
    "\n",
    "# TODO: How should I handle the no-rank ones?\n",
    "taxonomy = entrez.download_taxonomy(metadata.ncbi_taxonomy_id)\n",
    "cols = metadata.columns.union(taxonomy.columns)\n",
    "metadata = metadata.merge(taxonomy, left_on=['ncbi_taxonomy_id'], right_index=True, how='inner')[cols]\n",
    "metadata = metadata[~metadata.index.duplicated(keep='first')]\n",
    "metadata.to_csv(os.path.join(DATA_PATH, 'rna16s_metadata.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I might need to install gslm, here: https://github.com/ramanathanlab/genslm/blob/main/setup.cfg \n",
    "# Needed to install a Rust compiler before installing this. with this command: curl --proto '=https' --tlsv1.2 https://sh.rustup.rs -sSf | sh\n",
    "# After installing Rust, I also needed to run apt install pkg-config\n",
    "# Still didn't work, so I tried installing libssl-dev using apt. \n",
    "# Still didn't work, so I tried downgrading Python to 3.7. That finally worked. Not sure why, something with bugs when trying to install h5py and tokenizers. \n",
    "# Unfortunately, when I did this, I need to reinstall a ton of stuff. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of testing dataset: 587\n",
      "Size of training dataset: 2084\n",
      "Size of validation dataset: 465\n"
     ]
    }
   ],
   "source": [
    "# Seems to be a lot of downloaded genomes... what is the comparative size of the training and testing set?\n",
    "test_dataset = FeatureDataset(os.path.join(DATA_PATH, 'testing_datasets.h5'), feature_type='aa_1mer')\n",
    "train_dataset = FeatureDataset(os.path.join(DATA_PATH, 'training_datasets.h5'), feature_type='percent_oxygen_genes')\n",
    "val_dataset = FeatureDataset(os.path.join(DATA_PATH, 'validation_datasets.h5'), feature_type='aa_1mer')\n",
    "\n",
    "print('Size of testing dataset:', len(test_dataset))\n",
    "print('Size of training dataset:', len(train_dataset))\n",
    "print('Size of validation dataset:', len(val_dataset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>contig_size</th>\n",
       "      <th>Class</th>\n",
       "      <th>physiology</th>\n",
       "      <th>aa_1mer_prediction</th>\n",
       "      <th>aa_2mer_prediction</th>\n",
       "      <th>aa_3mer_prediction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_1</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_2</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_3</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_4</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_45</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_46</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_47</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_48</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_49</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16120 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            genome_id  contig_size         Class   physiology  \\\n",
       "id                                                                              \n",
       "GCF_000425825_2000_0    GCF_000425825         2000       Bacilli  Facultative   \n",
       "GCF_000425825_2000_1    GCF_000425825         2000       Bacilli  Facultative   \n",
       "GCF_000425825_2000_2    GCF_000425825         2000       Bacilli  Facultative   \n",
       "GCF_000425825_2000_3    GCF_000425825         2000       Bacilli  Facultative   \n",
       "GCF_000425825_2000_4    GCF_000425825         2000       Bacilli  Facultative   \n",
       "...                               ...          ...           ...          ...   \n",
       "GCF_000008665_40000_45  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_40000_46  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_40000_47  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_40000_48  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_40000_49  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "\n",
       "                       aa_1mer_prediction aa_2mer_prediction  \\\n",
       "id                                                             \n",
       "GCF_000425825_2000_0          facultative        facultative   \n",
       "GCF_000425825_2000_1          facultative        facultative   \n",
       "GCF_000425825_2000_2          facultative        facultative   \n",
       "GCF_000425825_2000_3          facultative        facultative   \n",
       "GCF_000425825_2000_4          facultative        facultative   \n",
       "...                                   ...                ...   \n",
       "GCF_000008665_40000_45           anaerobe           anaerobe   \n",
       "GCF_000008665_40000_46           anaerobe           anaerobe   \n",
       "GCF_000008665_40000_47           anaerobe           anaerobe   \n",
       "GCF_000008665_40000_48           anaerobe           anaerobe   \n",
       "GCF_000008665_40000_49           anaerobe           anaerobe   \n",
       "\n",
       "                       aa_3mer_prediction  \n",
       "id                                         \n",
       "GCF_000425825_2000_0          facultative  \n",
       "GCF_000425825_2000_1          facultative  \n",
       "GCF_000425825_2000_2          facultative  \n",
       "GCF_000425825_2000_3          facultative  \n",
       "GCF_000425825_2000_4          facultative  \n",
       "...                                   ...  \n",
       "GCF_000008665_40000_45           anaerobe  \n",
       "GCF_000008665_40000_46           anaerobe  \n",
       "GCF_000008665_40000_47           anaerobe  \n",
       "GCF_000008665_40000_48           anaerobe  \n",
       "GCF_000008665_40000_49           anaerobe  \n",
       "\n",
       "[16120 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigs_datasets = FeatureDataset(os.path.join(DATA_PATH, 'rna16s', 'testing_dataset.h5'), feature_type='aa_1mer')\n",
    "contigs_datasets.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>contig_size</th>\n",
       "      <th>Class</th>\n",
       "      <th>physiology</th>\n",
       "      <th>aa_1mer_prediction</th>\n",
       "      <th>aa_2mer_prediction</th>\n",
       "      <th>aa_3mer_prediction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_2000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>2000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_5000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>5000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_7000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>7000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_10000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>10000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000425825_20000_0</th>\n",
       "      <td>GCF_000425825</td>\n",
       "      <td>20000</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "      <td>facultative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_7000_0</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>7000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_10000_0</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>10000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_20000_0</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>20000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_30000_0</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>30000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008665_40000_0</th>\n",
       "      <td>GCF_000008665</td>\n",
       "      <td>40000</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "      <td>anaerobe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           genome_id  contig_size         Class   physiology  \\\n",
       "id                                                                             \n",
       "GCF_000425825_2000_0   GCF_000425825         2000       Bacilli  Facultative   \n",
       "GCF_000425825_5000_0   GCF_000425825         5000       Bacilli  Facultative   \n",
       "GCF_000425825_7000_0   GCF_000425825         7000       Bacilli  Facultative   \n",
       "GCF_000425825_10000_0  GCF_000425825        10000       Bacilli  Facultative   \n",
       "GCF_000425825_20000_0  GCF_000425825        20000       Bacilli  Facultative   \n",
       "...                              ...          ...           ...          ...   \n",
       "GCF_000008665_7000_0   GCF_000008665         7000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_10000_0  GCF_000008665        10000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_20000_0  GCF_000008665        20000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_30000_0  GCF_000008665        30000  Archaeoglobi     Anaerobe   \n",
       "GCF_000008665_40000_0  GCF_000008665        40000  Archaeoglobi     Anaerobe   \n",
       "\n",
       "                      aa_1mer_prediction aa_2mer_prediction aa_3mer_prediction  \n",
       "id                                                                              \n",
       "GCF_000425825_2000_0         facultative        facultative        facultative  \n",
       "GCF_000425825_5000_0         facultative        facultative        facultative  \n",
       "GCF_000425825_7000_0         facultative        facultative        facultative  \n",
       "GCF_000425825_10000_0        facultative        facultative        facultative  \n",
       "GCF_000425825_20000_0        facultative        facultative        facultative  \n",
       "...                                  ...                ...                ...  \n",
       "GCF_000008665_7000_0            anaerobe           anaerobe           anaerobe  \n",
       "GCF_000008665_10000_0           anaerobe           anaerobe           anaerobe  \n",
       "GCF_000008665_20000_0           anaerobe           anaerobe           anaerobe  \n",
       "GCF_000008665_30000_0           anaerobe           anaerobe           anaerobe  \n",
       "GCF_000008665_40000_0           anaerobe           anaerobe           anaerobe  \n",
       "\n",
       "[328 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigs_datasets.metadata.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "store = pd.HDFStore(os.path.join(DATA_PATH, 'rna16s', 'testing_dataset.h5'))\n",
    "# store.get_node('embedding.rna16s')._f_rename('embedding_rna16s')\n",
    "store.get_node('labels')._f_rename('metadata')\n",
    "store.close()\n",
    "\n",
    "store = pd.HDFStore(os.path.join(DATA_PATH, 'rna16s', 'training_dataset.h5'))\n",
    "# store.get_node('embedding.rna16s')._f_rename('embedding_rna16s')\n",
    "store.get_node('labels')._f_rename('metadata')\n",
    "store.close()\n",
    "\n",
    "store = pd.HDFStore(os.path.join(DATA_PATH, 'rna16s', 'validation_dataset.h5'))\n",
    "# store.get_node('embedding.rna16s')._f_rename('embedding_rna16s')\n",
    "store.get_node('labels')._f_rename('metadata')\n",
    "store.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gc_content', 'number_of_genes', 'aa_mean_num_carbon',\n",
       "       'aa_mean_num_oxygen', 'aa_mean_num_nitrogen', 'aa_mean_num_sulfur',\n",
       "       'cds_mean_num_carbon', 'cds_mean_num_oxygen', 'cds_mean_num_nitrogen'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_hdf(os.path.join(DATA_PATH, 'validation_datasets.h5'), key='chemical')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aerobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
